password-hash/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+    /// recommended default), password, salt, and optional [`Params`].
+    ///
+    /// Use [`Params::new`] or [`Params::default`] to use the default
+    /// parameters for a given algorithm.
+    fn hash_password(
+        &self,
+        alg: Option<Algorithm>,
+        password: &[u8],
+        salt: Salt,
+        params: Params,
+    ) -> Result<PasswordHash, PhfError>;
+}
+
+/// Password hash.
+///
+/// This type corresponds to the parsed representation of a PHC string as
+/// described in the [PHC string format specification][1].
+///
+/// PHC strings have the following format:
+///
+/// ```text
+/// $<id>[$<param>=<value>(,<param>=<value>)*][$<salt>[$<hash>]]
+/// ```
+///
+/// where:
+///
+/// - `<id>` is the symbolic name for the function
+/// - `<param>` is a parameter name
+/// - `<value>` is a parameter value
+/// - `<salt>` is an encoding of the salt
+/// - `<hash>` is an encoding of the hash output
+///
+/// The string is then the concatenation, in that order, of:
+///
+/// - a `$` sign;
+/// - the function symbolic name;
+/// - optionally, a `$` sign followed by one or several parameters, each with a `name=value` format;
+///   the parameters are separated by commas;
+/// - optionally, a `$` sign followed by the (encoded) salt value;
+/// - optionally, a `$` sign followed by the (encoded) hash output (the hash output may be present
+///   only if the salt is present).
+///
+/// [1]: https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md#specification
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PasswordHash {
+    /// Password hashing [`Algorithm`].
+    ///
+    /// This corresponds to the `<id>` field in a PHC string, a.k.a. the
+    /// symbolic name for the function.
+    pub algorithm: Algorithm,
+
+    /// Algorithm-specific [`Params`].
+    ///
+    /// This corresponds to the set of `$<param>=<value>(,<param>=<value>)*`
+    /// name/value pairs in a PHC string.
+    pub params: Params,
+
+    /// [`Salt`] string for personalizing a password hash output.
+    ///
+    /// This corresponds to the `<salt>` value in a PHC string.
+    pub salt: Option<Salt>,
+
+    /// Password hashing function [`Output`], a.k.a. hash/digest.
+    ///
+    /// This corresponds to the `<hash>` output in a PHC string.
+    pub hash: Option<Output>,
+}
+
+impl PasswordHash {
+    /// Generate a password hash using the supplied algorithm.
+    pub fn generate(
+        phf: impl PasswordHashingFunction,
+        password: impl AsRef<[u8]>,
+        salt: Salt,
+        params: Params,
+    ) -> Result<Self, PhfError> {
+        phf.hash_password(None, password.as_ref(), salt, params)
+    }
+
+    /// Verify this password hash using the specified set of supported
+    /// [`PasswordHashingFunction`] objects.
+    pub fn verify_password(
+        &self,
+        phfs: &[&dyn PasswordHashingFunction],
+        password: impl AsRef<[u8]>,
+    ) -> Result<(), VerifyError> {
```

Comment:
@newpavlov regarding this comment:

> How to make it easier support of several algorithms at once? 

That's implemented by this method. Prospectively it could work like:

```rust
password_hash.verify_password(&[Argon2, Scrypt], password)
```